tools/interactive_fit.py

```diff
@@ -0,0 +1,359 @@
+import functools
+
+
+class InteractiveFit2D(Fit):
+    """A class that provides a visual_guess method which provides
+    an graphical, interactive way of guessing initial fitting parameters."""
+    def _update_plot(self, val):
+        """Callbak to redraw the plot to reflect the new parameter values."""
+        # Since all sliders call this same callback without saying who they are
+        # I need to update the values for all parameters. This can be
+        # circumvented by creating a seperate callback function for each
+        # parameter.
+        for p in self.params:
+            # p.value = self.sliders[p].val?
+            self.params[self.params.index(p)].value = self.sliders[p].val
+        # Also see the comment below about using keyword arguments (line 57)
+        self.my_plot.set_ydata(self.vec_func(self.xpoints, *list(p.value for p in self.params)))
+
+    def visual_guess(self, n_points=50):
+        """Create a matplotlib window with sliders for all parameters
+        in this model, so that you may graphically guess initial fitting
+        parameters. n_points is the number of points drawn for the plot.
+        Data points are plotted as blue points, the proposed model as
+        a red line.
+        Slider extremes are taken from the parameters where possible. If
+        these are not provided, the minimum is 0; and the maximum is value*2.
+        If no initial value is provided, it defaults to 1."""
+
+        x_min = np.min(self.xdata)
+        x_max = np.max(self.xdata)
+        # It would be nice if I could get these from model someway...
+        y_min = 0
+        y_max = 1
+
+        if len(self.vars) != 1:
+            raise IndexError("Only supports 2D problems!")
+
+        ax = plt.subplot(111)
+        plt.subplots_adjust(bottom=0.1 + 0.05*len(self.params))
+        plt.title("Fit")
+
+        self.xpoints = np.linspace(x_min, x_max, n_points)
+
+        plt.ylim(y_min, y_max)  # Y-Domain for self.model
+        plt.xlim(x_min, x_max)  # X-domain for self.model
+
+        # self.vec_func should probably be replaced for self.model.
+        self.vec_func = lambdify(self.vars+self.params, self.model, "numpy")
+        # It might be better to use keyword arguments (p.name: p.value), but
+        # I don't know (any more) if that's supported by self.model.
+        vals = self.vec_func(self.xpoints, *list(p.value for p in self.params))
+        self.my_plot, = ax.plot(self.xpoints, vals, color='red')
+        ax.scatter(self.xdata, self.ydata, c='blue')
+        plt.axis([x_min, x_max, y_min, y_max])
+
+        i = 0.05
+        self.sliders = {}
+        for p in self.params:
+            if not p.fixed:
+                axbg = 'lightgoldenrodyellow'
+            else:
+                axbg = 'red'
+            # start-x, start-y, width, height
+            ax = plt.axes([0.162, i, 0.68, 0.03], axisbg=axbg)
+
+            if not hasattr(p, "value") or p.value is None:
```

Comment:
Value is 1.0 by default. And although I guess it's possible for some yerk to put p.value to None, I think you can just assume it to be a float or int. And you don't have to check if any of these attributes exist. They always do.